### DISCUSSION TITLE
`Either` over enums: which one is better?

### DISCUSSION THREAD

**Discussion by drHuangMHT** [libp2p/rust-libp2p#3902] [General]
I've been writing some macros to help compose connection handlers like [these](https://github.com/Owlput/OwlNest/blob/0.0.1/owlnest/src/macros.rs). Reason? Those binary-tree-like nested `Either`s are driving me crazy! The code seems to pass type checks but I'm not sure how well it works. Before digging deeper and deeper into macros, I have a question to ask: which one is better, generated nested `Either` or generated enums that bundles everything together? In my opinion, enums are much more intuitive, at least to me. And they are much easier to manage when you want to manually combine `NetworkBehaviour`s(or really, `ConnectionHandler`). So, will there be one day we kick nested `Either` out of our code?

**Comment by thomaseizinger** [comment]
My vision for this is to push forward on https://github.com/libp2p/rust-libp2p/issues/3591. That will remove the `Error` type from `ConnectionHandler` and simply make that 2nd type parameter on `SwarmEvent` go away completely :)

**Reply by drHuangMHT** [reply]
Sounds unrelated. May you give more details about it?

**Reply by thomaseizinger** [reply]
The only place where you encounter nested `Either`s in a typical rust-libp2p usage is when declaring the type of `SwarmEvent` with its 2nd type parameter for the " error" (i.e connection close reason) of the handler.

If that is not what you are hitting, can you explain in more detail when you are encountering nested `Either`s?

**Reply by drHuangMHT** [reply]
Other either-like types that combine two into one also count, for example `SelectUpgrade` and `ConnectionHandlerSelect`.

**Reply by thomaseizinger** [reply]
We'd need even more macros if we wanted to generate enums for them. Most of that complexity is hidden behind the `NetworkBehavior` derive macro except for what I mentioned above with the `ConnectionHandler`s `Error` type.

How come it surfaces so badly for you in other places?

**Reply by drHuangMHT** [reply]
> We'd need even more macros if we wanted to generate enums for them. Most of that complexity is hidden behind the `NetworkBehavior` derive macro except for what I mentioned above with the `ConnectionHandler`s `Error` type.
> 
> How come it surfaces so badly for you in other places?

when you have more than two subprotocols the tree structure will only be a bit annoying, but I still want to make it better. I want it to look like `tokio::select!`.
when handshaking with a unknown peer, it's best to leak as little information about yourself as possible instead of dumping everything, but I don't think it's possible with derive macro only.

**Reply by thomaseizinger** [reply]
> > We'd need even more macros if we wanted to generate enums for them. Most of that complexity is hidden behind the `NetworkBehavior` derive macro except for what I mentioned above with the `ConnectionHandler`s `Error` type.
> > How come it surfaces so badly for you in other places?
> 
> when you have more than two subprotocols the tree structure will only be a bit annoying, but I still want to make it better. I want it to look like `tokio::select!`.

Are you aware of `MultiHandler`? Typically, you'll have one `NetworkBehaviour` per protocol and not compose multiple `ConnectionHandler`s yourself.

>  when handshaking with a unknown peer, it's best to leak as little information about yourself as possible instead of dumping everything, but I don't think it's possible with derive macro only.

What do you mean with "leak"? If you don't want to allow connections to peers that you don't know, you could utilize `libp2p::allow_block_list`. No data will be sent to a peer that is blocked / not allowed (apart from the handshake itself to figure out who it is).

**Reply by drHuangMHT** [reply]
I guess only the code tells. Almost there to compose handlers and soon behaviours.

**Reply by drHuangMHT** [reply]
Now behaviours can be composed with the macro above, the code compiles but haven't been tested in practice.

**Reply by drHuangMHT** [reply]
@thomaseizinger The code now works as intended! If interested please have a look and try the macros with my application!

**Reply by thomaseizinger** [reply]
Can you link what specifically you want me to have a look at?

**Reply by drHuangMHT** [reply]
> Can you link what specifically you want me to have a look at?

same as above, https://github.com/Owlput/OwlNest/blob/0.0.1/owlnest/src/macros.rs

**Reply by thomaseizinger** [reply]
Sorry, I still don't follow why you need these macros. Can you show me some code that uses them? I looked through your repo but can't find anything.

**Reply by drHuangMHT** [reply]
> Sorry, I still don't follow why you need these macros. Can you show me some code that uses them? I looked through your repo but can't find anything.

`behaviour_select!` is used in https://github.com/Owlput/OwlNest/blob/0.0.1/owlnest/src/net/p2p/swarm/behaviour.rs at line 21. It is not that hard to compose behaviours but still involves `ConnectionHandlerSelect`, and if you have multiple different connection handlers for subprotocols you need to use `SelectUpgrade`, both of them will be heavily nested when number of cases grow.

**Reply by drHuangMHT** [reply]
However I do find one abnormality. The connection seems to be in constant negotiation because I see the iterator for producing `UpgradeInfo` being actively constructed and iterated over even after all negotiation succeeded and there appears to be no new inbound or outbound connection. For behaviours there are `kad`, `identify`, `mdns`, `relay_client` and `relay_server` with relayed transport used.

**Reply by thomaseizinger** [reply]
> However I do find one abnormality. The connection seems to be in constant negotiation because I see the iterator for producing `UpgradeInfo` being actively constructed and iterated over even after all negotiation succeeded and there appears to be no new inbound or outbound connection. For behaviours there are `kad`, `identify`, `mdns`, `relay_client` and `relay_server` with relayed transport used.

If you are already on latest master, then the reason for that is that we need to check whether the list of supported protocols changes on a connection.

**Reply by thomaseizinger** [reply]
> > Sorry, I still don't follow why you need these macros. Can you show me some code that uses them? I looked through your repo but can't find anything.
> 
> `behaviour_select!` is used in https://github.com/Owlput/OwlNest/blob/0.0.1/owlnest/src/net/p2p/swarm/behaviour.rs at line 21. It is not that hard to compose behaviours but still involves `ConnectionHandlerSelect`, and if you have multiple different connection handlers for subprotocols you need to use `SelectUpgrade`, both of them will be heavily nested when number of cases grow.

Are you aware of the `NetworkBehaviour` custom derive? It does all of the code-generation for you. You can use it by putting `#[derive(NetworkBehaviour)]` on a struct that contains multiple other behaviours.

**Reply by drHuangMHT** [reply]
> Are you aware of the `NetworkBehaviour` custom derive? It does all of the code-generation for you. You can use it by putting `#[derive(NetworkBehaviour)]` on a struct that contains multiple other behaviours.

Of course I am, it's right above line 21 and the derive macro will work just fine. You can't possibly miss that though?

**Reply by thomaseizinger** [reply]
> > Are you aware of the `NetworkBehaviour` custom derive? It does all of the code-generation for you. You can use it by putting `#[derive(NetworkBehaviour)]` on a struct that contains multiple other behaviours.
> 
> Of course I am, it's right above line 21 and the derive macro will work just fine.

Is there a reason you don't want to use it then? Or why go through the pain of writing `macro_rules` instead?

**Reply by drHuangMHT** [reply]
> > Of course I am, it's right above line 21 and the derive macro will work just fine.
> 
> Is there a reason you don't want to use it then? Or why go through the pain of writing `macro_rules` instead?

I don't have answer to that yet. But back to the question I raised for this discussion, is a nested generic either-like type better than dedicated enumeration? I will say yes because it is a very well defined interface, but as the number of items grows it becomes harder to maintain, at least in my mind.

**Reply by thomaseizinger** [reply]
> > > Of course I am, it's right above line 21 and the derive macro will work just fine.
> > 
> > Is there a reason you don't want to use it then? Or why go through the pain of writing `macro_rules` instead?
> 
> I don't have answer to that yet. But back to the question I raised for this discussion, is a nested generic either-like type better than dedicated enumeration? I will say yes because it is a very well defined interface, but as the number of items grows it becomes harder to maintain, at least in my mind.

Another important aspect is that a type-based composition approach (at least for low-level components) is preferable because it doesn't need macros. `ConnectionHandlerSelect` is just regular Rust.

When done well, this doesn't surface to the user. Currently it leaks to the user through the `SwarmEvent` as described above.

Whenever you generate code, it should ideally be something that the user never interacts with (like the `Serialize` impl of `serde`).

That is an aspect of our solution that I don't like. The user needs to interact with an enum that they can't see.

**Reply by drHuangMHT** [reply]
> Another important aspect is that a type-based composition approach (at least for low-level components) is preferable because it doesn't need macros. `ConnectionHandlerSelect` is just regular Rust.

That makes sense. Case closed for now.

> When done well, this doesn't surface to the user. Currently it leaks to the user through the `SwarmEvent` as described above.

That's indeed a major concern because it is so annoying that I have to make a type alias.
